File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.py
from collections import defaultdict

def solution(genres, plays):
    dict_genres = defaultdict(list)
    for idx, (genre, play) in enumerate(zip(genres, plays)):  # 1~10000
        dict_genres[genre].append((idx, play))
        
    plays_genres = list(dict_genres.values())
    for plays in plays_genres:    # 0~99
        plays.sort(key=lambda x:[-x[1], x[0]])    # 10000log10000
        plays.append(sum(play for _, play in plays))
    plays_genres.sort(key=lambda x:-x[-1])  #100log100
    
    answer = []
    for plays in plays_genres:    # 100
        for idx, play in enumerate(plays):
            if isinstance(play, tuple) and idx < 2:
                answer.append(play[0])
    return answer

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.java
import java.util.*;

class Music {
    String genre;
    int genrePlay;
    int play;
    int no;
    
    public Music(String genre, int genrePlay, int play, int no) {
        this.genre = genre;
        this.genrePlay = genrePlay;
        this.play = play;
        this.no = no;
    }
    
    public String toString() {
        return genre + no;
    }
}

class Solution {
    public List<Integer> solution(String[] genres, int[] plays) {
        int n = genres.length;

        // 10,000 : 장르별 합계 구하기
        Map<String, Integer> genrePlayMap = new HashMap<>();
        for (int i=0; i<n; i++) {
            String genre = genres[i];
            if (genrePlayMap.containsKey(genre)) {
                genrePlayMap.put(genre, genrePlayMap.get(genre)+plays[i]);
            } else {
                genrePlayMap.put(genre, plays[i]);
            }
        }
        
        // 10,000 : 객체 만들기
        Music[] musics = new Music[n];
        for (int i=0; i<n; i++) {
            musics[i] = new Music(genres[i], genrePlayMap.get(genres[i]), plays[i], i);
        }
        
        // 10,000log10,000 : 우선순위 기준 정렬
        Arrays.sort(musics, (a, b)->{
            if (a.genrePlay != b.genrePlay) return b.genrePlay - a.genrePlay;
            if (a.play != b.play) return b.play - a.play;
            return a.no - b.no;
        });
        // System.out.println(Arrays.toString(musics));
        
        // 10,000 : 장르별 2개
        Map<String, Integer> pickedCount = new HashMap<>();
        List<Integer> answer = new ArrayList<>();
        for (Music m : musics) {
            int cnt = pickedCount.getOrDefault(m.genre, 0);
            if (cnt < 2) {
                answer.add(m.no);
                pickedCount.put(m.genre, cnt + 1);
            }
        }
        
        return answer;
    }
}